### PR TITLE
S1 refactor + Readout bug fix

### PR DESCRIPTION
**Pull request by petergaemers**
This pull request:

- Makes all s1 functions callable from the outside. So now they're all staticmethod so if you want to calculate some photon arrival timings yourself this is now easy.
- Fix readout bug. The last part of the data was not read out from the buffer.
- Make raw_optical slightly nicer

Side remark: Tests actually work when run locally. Through github we have issues with the private repo

### CODE REVIEW

File: wfsim/core.py [hunk 1]
```diff
@@ -291,89 +291,105 @@ def __call__(self, instruction):
...
+        #TODO Looks nice but don't know if this actually works
         try:
-            self._photon_timings = np.append(self._photon_timings,
-                t + getattr(self, recoil_type.lower())(n_photons))
+            recoil_types = [k for k in vars(NestId) if not k.startswith('_')]
+            for recoil in recoil_types:
+                if recoil in NestId.recoil:
+                    _photon_timings = t + getattr(S1, recoil_type.lower())(n_photons=n_photons,
+                                                                           config=config,
+                                                                           singlet_triplet_delays=Pulse.singlet_triplet_delays)
```
Comment (zhut19) [inline wfsim/core.py:351]: To use this we also need to pass the 'phase, singlet decay time, triplet decay time' as arguments.

Comment (petergaemers) [reply wfsim/core.py]: phase is added, lifetimes are in the config

**Comment by ramirezdiego** [comment]
@petergaemers Please make sure I didn't resolve conflicts in the wrong direction, in `strax_interface.py`.

File: tests/test_wfsim.py [hunk 2]
```diff
@@ -4,27 +4,16 @@
...
-        st = strax.Context(
-            storage=tempdir,
-            config=dict(
-                nchunk=1, event_rate=1, chunk_size=10,
-                detector='XENON1T',
-                fax_config='https://raw.githubusercontent.com/XENONnT/strax_auxiliary_files/7ba4875b52162cbbe9284faf66a6b9193a254a30/sim_files/fax_config_1t.json',  # noqa
-                **straxen.contexts.x1t_common_config),
-            **straxen.contexts.common_opts)
-        st.register(wfsim.RawRecordsFromFax1T)
+        st = straxen.contexts.xenon1t_simulation(output_folder=tempdir)
```
Comment (HenningSE) [inline tests/test_wfsim.py:16]: The function will now use the private `fax_config` file defined in https://github.com/XENONnT/straxen/blob/master/straxen/contexts.py#L329-L340 or? The docstring outlines that it should work with publicly available files.

File: tests/test_wfsim.py [hunk 3]
```diff
@@ -39,16 +28,7 @@ def test_sim_nT():
...
-            storage=tempdir,
-            config=dict(
-                nchunk=1, event_rate=1, chunk_size=10,
-                detector='XENONnT',
-                fax_config='fax_config_nt.json',
-                **straxen.contexts.xnt_common_config),
-            **straxen.contexts.common_opts)
-        st.register(wfsim.RawRecordsFromFaxNT)
-
+        st = straxen.contexts.xenonnt_simulation(output_folder=tempdir)
```
Comment (HenningSE) [inline tests/test_wfsim.py:31]: Same as for `test_sim_1T`.

File: wfsim/core.py [hunk 4]
```diff
@@ -291,89 +291,106 @@ def __call__(self, instruction):
...
+            _photon_timings = t + np.random.exponential(config['s1_decay_time'], n_photons)
+            _photon_timings += np.random.normal(0,config['s1_decay_spread'],len(_photon_timings))
+            return _photon_timings
 
-        #TODO Fix the other recoil types to refer to the numbers
+        #TODO Looks nice but don't know if this actually works
         try:
-            self._photon_timings = np.append(self._photon_timings,
-                t + getattr(self, recoil_type.lower())(n_photons))
+            recoil_types = [k for k in vars(NestId) if not k.startswith('_')]
```
Comment (HenningSE) [inline wfsim/core.py:346]: Here you want to look over `ER`, `NR` and `ALPHA` and **not** over `_ALL` right?

File: wfsim/core.py [hunk 5]
```diff
@@ -291,89 +291,106 @@ def __call__(self, instruction):
...
+            ly *= config['s1_detection_efficiency']
+        n_photons = np.random.binomial(n=n_photons, p=ly)
+        return n_photons
 
-        # print(self.config)
-        channels = np.arange(self.config['n_tpc_pmts'])#+1 for the channel map
-        p_per_channel = self.resource.s1_pattern_map(points)
-        p_per_channel[:, np.in1d(channels, self.turned_off_pmts)] = 0
+    @staticmethod
+    def photon_channels(points, n_photons,config,s1_pattern_map):
```
Comment (HenningSE) [inline wfsim/core.py:317]: Just to make the function definition consistent with the other functions, a space should be added in front of `config ` and `s1_pattern_map `.

File: wfsim/core.py [hunk 6]
```diff
@@ -291,89 +291,106 @@ def __call__(self, instruction):
...
+            recoil_types = [k for k in vars(NestId) if not k.startswith('_')]
+            for recoil in recoil_types:
+                if recoil in NestId.recoil:
+                    _photon_timings = t + getattr(S1, recoil_type.lower())(n_photons=n_photons,
+                                                                           config=config,
+                                                                           singlet_triplet_delays=Pulse.singlet_triplet_delays,
+                                                                           phase=self.phase)
+
         except AttributeError:
             raise AttributeError('Recoil type must be ER, NR, alpha or LED, not %s' % recoil_type)
```
Comment (HenningSE) [inline wfsim/core.py:355]: This Error message could maybe be updated to match the NEST IDs.

File: wfsim/core.py [hunk 7]
```diff
@@ -291,89 +291,106 @@ def __call__(self, instruction):
...
-    def alpha(self, size):
+    @staticmethod
+    def alpha(size,config,singlet_triplet_delays, phase):
         # Neglible recombination time
-        return self.singlet_triplet_delays(size, self.config['s1_ER_alpha_singlet_fraction'])
+        return singlet_triplet_delays(size, config['s1_ER_alpha_singlet_fraction'],config, phase)
 
-    def led(self, size):
+    @staticmethod
+    def led(size,config, **kwargs):
```
Comment (HenningSE) [inline wfsim/core.py:363]: Why do we need the `kwargs`?

File: wfsim/core.py [hunk 8]
```diff
@@ -1136,7 +1149,7 @@ def get_truth(self, instruction, truth_buffer):
             n_dpe_bot = getattr(pulse, '_n_double_pe_bot', 0)
         tb['n_photon'] += n_dpe
         tb['n_photon'] -= np.sum(np.isin(channels, getattr(pulse, 'turned_off_pmts', [])))
-
+        #TODO this turned_off guy, check how this works with a config['turned_off_guys']
```
Comment (HenningSE) [inline wfsim/core.py:1152]: guy = pmt? :-)

File: wfsim/raw_optical.py [hunk 9]
```diff
@@ -74,116 +79,44 @@ def __call__(self, instructions, channels, timings, truth_buffer=None, **kwargs)
             # D) Run all clusters before the current source
             stop_at_this_group = False
             for ibqs in instb_queue:
-                for ptype in [1, 2]:  # S1 S2 PI Gate
+                for ptype in [1, 2, 4, 6]: # S1 S2 PI Gate
```
Comment (HenningSE) [inline wfsim/raw_optical.py:82]: Stupid question, but what are PI and Gate?

File: wfsim/raw_optical.py [hunk 10]
```diff
@@ -74,116 +79,44 @@ def __call__(self, instructions, channels, timings, truth_buffer=None, **kwargs)
...
+                if stop_at_this_group: 
+                    break
+                self.digitize_pulse_cache() # from pulse cache to raw data
                 yield from self.ZLE()
-
+                
             self.source_finished = len(inst_queue) == 0 and np.sum(instb_filled) == 0
         pbar.close()
 
-    def sim_data(self, instruction, channels, timings):
```
Comment (HenningSE) [inline wfsim/raw_optical.py:109]: Now RawData.sim_data is used?

File: wfsim/strax_interface.py [hunk 11]
```diff
@@ -275,9 +275,14 @@ def __init__(self, config):
     strax.Option('fax_config',
                  default='https://raw.githubusercontent.com/XENONnT/private_nt_aux_files/master/sim_files/fax_config_nt.json?token=AHCU5AZMPZABYSGVRLDACR3ABAZUA'),
     strax.Option('gain_model',
+<<<<<<< Updated upstream
```
Comment (HenningSE) [inline wfsim/strax_interface.py]: Should be removed I think.

File: wfsim/strax_interface.py [hunk 12]
```diff
@@ -275,9 +275,14 @@ def __init__(self, config):
...
                  default=('to_pe_per_run', 'https://github.com/XENONnT/private_nt_aux_files/blob/master/sim_files/to_pe_nt.npy?raw=true'),
                  help='PMT gain model. Specify as (model_type, model_config).'),
     strax.Option('detector', default='xenonnt_detector', track=True),
+=======
```
Comment (HenningSE) [inline wfsim/strax_interface.py]: This one as well.

File: wfsim/strax_interface.py [hunk 13]
```diff
@@ -275,9 +275,14 @@ def __init__(self, config):
...
+                 help='PMT gain model. Specify as (model_type, model_config). Provided by context (Eith cmt or a constant)'),
+    strax.Option('detector', default='XENONnT', track=True),
+>>>>>>> Stashed changes
```
Comment (HenningSE) [inline wfsim/strax_interface.py]: and here.

**Review by HenningSE — COMMENTED** [review]
Hi Peter, thx for the PR. I think it's easier to understand the functions now that they are "standalone" functions in some cases.

File: tests/test_wfsim.py (hunk 2 above)
Comment (jorana) [reply tests/test_wfsim.py:16]: I think the answer is given here:
https://github.com/XENONnT/WFSim/pull/88

File: wfsim/core.py (hunk 4 above)
Comment (petergaemers) [reply wfsim/core.py:346]: yes

File: wfsim/core.py (hunk 7 above)
Comment (petergaemers) [reply wfsim/core.py:363]: Because it allows you to call it in line 350

File: wfsim/core.py (hunk 8 above)
Comment (petergaemers) [reply wfsim/core.py:1152]: pulse no longer had an attribute called "turned_off_pmts"

File: wfsim/raw_optical.py (hunk 9 above)
Comment (petergaemers) [reply wfsim/raw_optical.py:82]: Afterpulses

File: wfsim/raw_optical.py (hunk 10 above)
Comment (petergaemers) [reply wfsim/raw_optical.py:109]: Yes

**Pull request merged**